cuDSS/simple_matrix_market/matrix_market_reader.h

```diff
@@ -0,0 +1,197 @@
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <algorithm>
+
+enum MtxReaderStatus {
+    MtxReaderSuccess,
+    MtxReaderErrorFileNotFound,
+    MtxReaderErrorFileMemAllocFailed,
+    MtxReaderErrorWrongNnz,
+    MtxReaderErrorUpperViewButLowerFound,
+    MtxReaderErrorLowerViewButUpperFound,
+    MtxReaderErrorOutOfBoundRowIndex,
+    MtxReaderErrorOfBoundColIndex
+
+};
+
+template <typename _TYPE_>
+int matrix_reader(std::string filename, int& n, int& nnz, int** csr_offsets_h,
+                  int** csr_columns_h, _TYPE_** csr_values_h, cudssMatrixViewType_t mview, bool verbose=true, bool verbose_err = true)
+{
+    std::ifstream file(filename);
+    if (!file.is_open()) {
+        if (verbose_err)
+            std::cerr << "\033[38;5;196m Error: Could not open file \033[0m" << filename << std::endl;
+        // ! Do a mock malloc anyway so that subsequent free does not fail !
+        *csr_offsets_h = (int*)malloc(1);
+        *csr_columns_h = (int*)malloc(1);
+        *csr_values_h = (_TYPE_*)malloc(1);
```

Comment:
I would suggest to set them to `NULL` in order to make it clear that an error happened. For the subsequent free, you can add a simple check: `if (ptr) free(ptr);`